src/libraries/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryFormatterEventSource.cs

```diff
@@ -0,0 +1,158 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.Threading;
+
+namespace System.Runtime.Serialization.Formatters.Binary
+{
+    [EventSource(
+        Name = "System.Runtime.Serialization.Formatters.Binary.BinaryFormatterEventSource",
+        LocalizationResources = "FxResources.System.Runtime.Serialization.Formatters.SR")]
+    internal sealed class BinaryFormatterEventSource : EventSource
+    {
+        private const int EventId_SerializationStarted = 10;
+        private const int EventId_SerializationEnded = 11;
+        private const int EventId_SerializingObject = 12;
+        private const int EventId_DeserializationStarted = 20;
+        private const int EventId_DeserializationEnded = 21;
+        private const int EventId_DeserializingObject = 22;
+
+        // Used to keep track of whether a write operation is in progress. It's
```

Comment:
Should this rather be a problem for the listener to worry about?

I do not remember seeing a recursion protection code like this in any other EventSource. For example. it is plausible for networking event source to run into same problem, but I do not see any code related to it.

Reply:
Honestly I was torn on this. The reason I included this check is that I perceived it to be likely that an EventListener implementation might want to log events to disk. And one fairly common way of logging data to disk would be to call `File.OpenWrite` when the EventListener is constructed, then `formatter.Serialize(fileStream, new EventData() { /* ... */ })` over and over again until the listener is disposed.

To be perfectly honest, I don't have evidence that anybody is actually doing this. referencesource.microsoft.com isn't showing any `EventListener`-derived types at all, so I can't check their implementations. But it seemed wholly within the realm of possibility.

Reply:
We only have the intrinsic event listener that logs to ETW/LTTNG/EventPipe inbox. `EventListener`s are always 3rd party code, there is not that many of them and they have to implemented by experts to avoid problems like this.

Reply:
I'll remove the checks.